### PR TITLE
HS-1363: cleanup all the streams on disconnect.

### PR DESCRIPTION
**Pull request by artnaseef**
## Jira link(s)
- https://issues.opennms.org/browse/HS-1363

## Checklist
* [x] Follows Horizon Stream's [development guidelines.](https://github.com/OpenNMS/horizon-stream/wiki/Development-Guidelines)
* [x] Appropriate reviewer(s) have been selected.
* [x] Jira issue(s) have been updated to "In Review".
* [x] Includes [appropriate tests.](https://github.com/OpenNMS/horizon-stream/wiki/Test-Strategy)
* [x] Documentation has been updated as necessary.
* [x] Notify devops of changes to the Charts

### CODE REVIEW

File: minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/MinionGrpcClient.java [hunk 1]
```diff
@@ -300,8 +336,14 @@ private void initializeCloudReceiver() {
             .setLocation(ipcIdentity.getLocation())
             .setSystemId(ipcIdentity.getId())
             .build();
-        asyncStub.cloudToMinionMessages(identity, new CloudMessageObserver(cloudMessageHandler));
-        LOG.info("Initialized cloud receiver stream");
+
+        cloudToMinionStreamCancellableContext = Context.current().withCancellation();
+        cloudToMinionStreamCancellableContext.run(
+            () -> {
+                asyncStub.cloudToMinionMessages(identity, new CloudMessageObserver(cloudMessageHandler));
```
Comment (splatch) [inline minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/MinionGrpcClient.java:343]: I think code needs to be wrapped into condition cause context can be cancelled before runnable is executed.

```suggestion
                if (!Context.current().isCancelled()) {
                    asyncStub.cloudToMinionMessages(identity, new CloudMessageObserver(cloudMessageHandler));
                }
```

File: minion/minion-grpc/grpc-client/src/test/java/org/opennms/horizon/minion/grpc/MinionGrpcClientTest.java [hunk 2]
```diff
@@ -108,6 +128,74 @@ void testStartTls() throws Exception {
             target.setTlsEnabled(true);
             target.start();
 
+            //
+            // Verify the Results
+            //
+            Predicate<LogEvent> matcher1 =
+                (logEvent) ->
+                    (
+                        Objects.equals("TLS enabled for gRPC", logEvent.getMessage() )
```
Comment (splatch) [inline minion/minion-grpc/grpc-client/src/test/java/org/opennms/horizon/minion/grpc/MinionGrpcClientTest.java:137]: What this test is attempting to prove? It is better to verify interactions with mocked components which prove behavior of code under test. Log output which can be disabled at any time is not a reliable behavior verification.

**Review by splatch — APPROVED** [review]

File: minion/minion-grpc/grpc-client/src/test/java/org/opennms/horizon/minion/grpc/MinionGrpcClientTest.java (hunk 2 above)
Comment (artnaseef) [reply minion/minion-grpc/grpc-client/src/test/java/org/opennms/horizon/minion/grpc/MinionGrpcClientTest.java:137]: The verification is ensuring the TLS path is hit.  Unfortunately `channelBuilder` cannot be mocked, so it's not possible to verify that way.

File: minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/MinionGrpcClient.java (hunk 1 above)
Comment (artnaseef) [reply minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/MinionGrpcClient.java:343]: Do you have more information?  I found this snippet recommended online and followed that pattern.

Comment (artnaseef) [reply minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/MinionGrpcClient.java:343]: Looking at that carefullly, what happens if we don't check `Context.current().isCancelled()` first?  Also, is there something we should be doing when it is cancelled?

Comment (artnaseef) [reply minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/MinionGrpcClient.java:343]: Oh it seems there's another concern that creates; from the javadocs:

`Create a new context which is independently cancellable and also cascades cancellation from its parent. Callers must ensure that either [Context.CancellableContext.cancel(Throwable)](https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.CancellableContext.html#cancel-java.lang.Throwable-) or [Context.CancellableContext.detachAndCancel(Context, Throwable)](https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.CancellableContext.html#detachAndCancel-io.grpc.Context-java.lang.Throwable-) are called at a later point, in order to allow this context to be garbage collected.`

Comment (splatch) [reply minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/MinionGrpcClient.java:343]: It is indeed note from javadoc and example of how to use cancellable context.  I suppose that calling `cancel(null)` when we shutdown channel or entire client is fine.

Comment (artnaseef) [reply minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/MinionGrpcClient.java:343]: Good news - we do call cancel on it normally, so we should not have a leak.  The `handleDisconnect` method ensures that.

**Pull request merged**